### PR TITLE
Add Publish permission & action

### PR DESCRIPTION
**Pull request by Convly**
Added the "Publish" action/permissions

### CODE REVIEW

File: packages/strapi-admin/domain/action.js [hunk 1]
```diff
@@ -10,24 +10,29 @@ const actionFields = [
   'pluginName',
   'subjects',
   'conditions',
+  'options',
 ];
 
+const defaultAction = {
+  options: {
+    fieldsGranularity: true,
```
Comment (Convly) [inline packages/strapi-admin/domain/action.js:18]: Not sure about the naming for this one.
Do you have a preference @alexandrebodin?

File: packages/strapi-admin/services/content-type.js [hunk 2]
```diff
@@ -141,13 +142,13 @@ const getPermissionsWithNestedFields = (actions, { nestingLevel, restrictedSubje
...
-const cleanPermissionFields = (permissions, { nestingLevel }) =>
+const cleanPermissionFields = (permissions, { nestingLevel } = {}) =>
   permissions.map(perm => {
     const { action: actionId, fields, subject } = perm;
     const action = strapi.admin.services.permission.actionProvider.getByActionId(actionId);
     let newFields = fields;
 
-    if (!action.options.fieldsGranularity) {
+    if (!actionDomain.hasFieldsRestriction(action)) {
       newFields = null;
```
Comment (alexandrebodin) [inline packages/strapi-admin/services/content-type.js:152]: I think you can directly return here

File: packages/strapi-admin/validation/common-validators.js [hunk 3]
```diff
@@ -94,9 +102,9 @@ const updatePermissions = yup
             checkNoDuplicatedPermissions
           )
           .test(
-            'delete-fields-are-null',
-            'The action "plugins::content-manager.explorer.delete" must have fields set to null or undefined',
-            checkCTPermsDeleteHaveFieldsToNull
+            'fields-restriction',
+            'The actions must have fields set to null or undefined',
+            checkContentTypesFieldsRestriction
```
Comment (alexandrebodin) [inline packages/strapi-admin/validation/common-validators.js:107]: I think you should use the this.createError() of the test function to set a better message with the action name in it.

File: packages/strapi-admin/services/role.js [hunk 4]
```diff
@@ -209,22 +210,26 @@ const createRolesIfNoneExist = async ({ createPermissionsForAdmin = false } = {}
...
-      fieldsNullFor: ['plugins::content-manager.explorer.delete'],
       restrictedSubjects: ['plugins::users-permissions.user'],
     }
   );
 
-  const authorPermissions = editorPermissions.map(p => ({
-    ...p,
-    conditions: ['admin::is-creator'],
-  }));
+  const addIsCreatorCondition = map(p => _.merge({}, p, { conditions: ['admin::is-creator'] }));
```
Comment (alexandrebodin) [inline packages/strapi-admin/services/role.js:224]: if you go with some fp then go with

```js
 const authorPermissions = editorPermissions
  .filter(({ action }) => action !== ACTIONS.publish)
  .map(_.set('conditions', ['admin::is-creator']))
```

File: packages/strapi-admin/validation/common-validators.js [hunk 5]
```diff
@@ -85,20 +80,30 @@ const updatePermissions = yup
                 'field-nested',
                 'Fields format are incorrect (duplicates).',
                 checkFieldsDontHaveDuplicates
+              )
+              .test(
+                'fields-restriction',
+                'The permission at ${path} must have fields set to null or undefined',
+                function(fields) {
```
Comment (alexandrebodin) [inline packages/strapi-admin/validation/common-validators.js:87]: move function outside

**Pull request merged**